### DISCUSSION TITLE
video call

### DISCUSSION THREAD

**Discussion by GeorgeS2019** [SakulFlee/Godot-WebRTC-Match-Maker#153] [General]
Would be great if **local** video streaming starts before peers join.

![image](https://github.com/SakulFlee/Godot-WebRTC-Match-Maker/assets/49812372/d4d58def-95e8-41d9-8394-4f6d1bc98316)

![image](https://github.com/SakulFlee/Godot-WebRTC-Match-Maker/assets/49812372/930e35c7-d1bb-44db-874b-63e7000b6c42)

**Comment by SakulFlee** [comment]
1. Local video streaming starts as soon as a capture device is selected (i.e. webcam).  
That has nothing to do with peers connecting.
Add a config or choose the default device or something like it and you'll have video streaming immediately.

2. If the connection or debug panel is bothering you: Remove it! :)
Those two are modular and can be removed by deleting them from Godot and removing a single line in the `_Ready()` function of the demo.

3. The error you are getting is from the Match Maker configuration not having the slot configured.
The configuration file is **only renewed if it doesn't exist**.
You probably have run the Match Maker before, on an earlier version, where this slot didn't exist yet.

When starting the Match Maker server it prints the config location into console.
Either delete that file and restart the server (== will regenerate a new config _with_ the slot added by default), or, go in that file and add the `VideoCall` to it's slots like so:

```toml
(...)
[slots]
VideoCall = 2
```

@GeorgeS2019

**Reply by GeorgeS2019** [reply]
I have an external webcam recognized but video stream not displayed. the integrated camera is ok. suggestion how to configure

**Reply by SakulFlee** [reply]
MHH... That's weird. It works for me.

1. Are there any errors in Godot?
(Check the Debug error panel, not just the log)

2. Try selecting a different webcam if you have.
I had two listed, but only one physical webcam is connected. One would work, the other one wouldn't send any data.
Additionally, virtual webcams like OBS wouldn't show up, they seem to be ignored by the library.

3. Are you talking about getting your webcam to work **at all**, or, **on both sessions**?
AFAIK, Windows has a restriction in place that only allows one app accessing a physical webcam at a time.
I.e. you can only have one session with a local and one session with a remote stream, but not bi-directional (on Windows, with only one physical webcam).

4. Make sure your webcam actually works. Laptops often block access to it.
Windows had an integrated Camera app for example or check it with something like Discord (in the settings! No call required).

**Reply by SakulFlee** [reply]
If all of those points aren't helpful, I wonder if your webcam outputs some strange format that Godot doesn't support.

YUV (the standard), RGB, PNG and JPG are supported by Godot. Other formats may or may not work.

If you can, even if there is no error at all, please upload logs of a full connect.
I.e. opening the app -> selecting video demo -> selecting a capture device
Repeat for any capture device shown.

**Comment by GeorgeS2019** [comment]
This code display external webcam
[FlashCap.WindowsForms](https://github.com/kekyo/FlashCap/blob/854186a445e494548480ae0d6b607c1bf009eb57/samples/FlashCap.WindowsForms/MainForm.cs#L56)

```cs

            var characteristics = descriptor0.Characteristics.
                FirstOrDefault(c => c.PixelFormat != PixelFormats.Unknown);
```


https://github.com/SakulFlee/Godot-WebRTC-Match-Maker/blob/d027bc1c85059bf31cfc7d63567308983e3f7844/Godot%20Project/Demos/VideoCall/VideoCall.cs#L112
 ```cs
	private async void OnCaptureDeviceSelected(int index)
	{
		if (captureDevice != null)
		{
			await captureDevice.StopAsync();
		}

		var selectedCaptureDevice = captureDevices.EnumerateDescriptors().ElementAt(index);
		GD.Print($"[VideoCall] Device selected: {selectedCaptureDevice}");

		characteristics = selectedCaptureDevice.Characteristics[0];
		GD.Print($"[VideoCall] Characteristics chosen: {characteristics}");

		captureDevice = await selectedCaptureDevice.OpenAsync(
			characteristics,
			OnNewFrame
		);
		await captureDevice.StartAsync();
		GD.Print("[VideoCall] Capture device opened and started!");
	}
```

**Reply by SakulFlee** [reply]
I don't know what you are trying to say unfortunately. Please try to explain why this code is relevant?
The top code is from their WindowsForm module which we aren't even using.

Secondly, I am still waiting for logs from you to fix this. Without knowing what exactly is going on, I can't fix it for you.

Also, if you know what is wrong already, fix it and open a PR please :)
